java-engine/src/main/java/io/getunleash/engine/WasmInterface.java

```diff
@@ -136,6 +140,14 @@ public void takeState(int enginePtr, byte[] messageBytes) {
     }
   }
 
+  @Override
+  public String getState(int enginePtr) {
+    synchronized (engineLock) {
+      int ptr = (int) getState.apply(enginePtr)[0];
+      return instance.memory().readCString(ptr);
```

Comment:
I'm pretty sure this needs a free. I don't think just reading the string is enough to free it (if you were looking at readLog, that's intentionally not freed this side because it's freed on the Rust side).

I think you'd need a new function for this on the Rust side. Something like this

```
pub extern "C" fn free_cstring(ptr: *mut c_char) {
    if !ptr.is_null() {
        unsafe {
            let _ = CString::from_raw(ptr);
        }
    }
}
```

The FFI layer should have a function called free_response which should do what you want.